src/main/clojure/clara/rules/engine.cljc

```diff
@@ -1748,50 +1748,74 @@
         (when (flush-updates *current-session*)
           (recur (mem/next-activation-group transient-memory) next-group))))))
 
-(deftype LocalSession [rulebase memory transport listener get-alphas-fn]
+(deftype LocalSession [rulebase memory transport listener get-alphas-fn pending-operations]
   ISession
   (insert [session facts]
-    (let [transient-memory (mem/to-transient memory)
-          transient-listener (l/to-transient listener)]
 
-      (l/insert-facts! transient-listener facts)
-
-      (binding [*pending-external-retractions* (atom [])]
-        ;; Bind the external retractions cache so that any logical retractions as a result
-        ;; of these insertions can be cached and executed as a batch instead of eagerly realizing
-        ;; them.  An external insertion of a fact that matches
-        ;; a negation or accumulator condition can cause logical retractions.
-        (doseq [[alpha-roots fact-group] (get-alphas-fn facts)
-                root alpha-roots]
-          (alpha-activate root fact-group transient-memory transport transient-listener))
-        (external-retract-loop get-alphas-fn transient-memory transport transient-listener))
+    (let [new-pending-operations (conj pending-operations (->PendingUpdate :insertion (if (coll? facts)
+                                                                                        facts
+                                                                                        (into [] facts))))]
```

Comment:
You are correct that I'm defending against a mutable collection here.  Currently if a user passes in a mutable collection and then mutates it later it doesn't matter since Clara will insert/retract the facts in it before returning from the insert/retract call.  Now that this is delayed mutation of the collection would change the behavior unless we specifically do something to be passive.  Note that we only do this if (coll? facts) returns false.  From the [docs](https://clojuredocs.org/clojure.core/coll_q) this should return true for any Clojure persistent collection.  Basically, we'd only incur the copying costs if a user passes a non-Clojure collection; otherwise we'd just incur what [amounts to an instance check](https://github.com/clojure/clojure/blob/clojure-1.8.0/src/clj/clojure/core.clj#L5977).  Particularly since Clara has a [top-level Java API](https://github.com/cerner/clara-rules/blob/0.13.0/src/main/clojure/clara/rules/java.clj) being passive to Java usage seems desirable.